### PR TITLE
Fix EngineerRepair stances and make the trait more customisable.

### PR DESCRIPTION
**Pull request by MustaphaTR**
Engineer Repair stances didn't work for 2 reasons. In the order generator it was not actually checking info.ValidStances, but directly if the target is Ally or not. 2 the Stance check in the activity was wrong way around.

This PR in addition to that:
- Adds Types: to EngineerRepair and EngineerRepairable. Both being empty is also a valid match.
- Made cusors definable in yaml, it was hardcoded to `goldwrench` and `goldwrench-blocked`.
- Made EngineerRepair conditional, tho not EngineerRepairable.
- Added RepairSounds to EngineerRepair. In RA2 repairing structures played a sound.

TESTCASE is only for stance fix. I didn't set up for anything else.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/RepairBuilding.cs [hunk 1]
```diff
@@ -9,24 +9,23 @@
...
 using OpenRA.Primitives;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	class RepairBuilding : Enter
 	{
-		readonly EnterBehaviour enterBehaviour;
-		readonly Stance validStances;
+		EngineerRepairInfo info;
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/RepairBuilding.cs]: Please make this `readonly`.

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 2]
```diff
@@ -18,8 +19,11 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Can instantly repair other actors, but gets consumed afterwards.")]
-	class EngineerRepairInfo : ITraitInfo
+	class EngineerRepairInfo : ConditionalTraitInfo
 	{
+		[Desc("Uses the \"EngineerRepairable\" trait to determine whether or not we should clone a produced unit.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: ...clone a produced unit?

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 3]
```diff
@@ -29,21 +33,32 @@ class EngineerRepairInfo : ITraitInfo
...
 
-		public object Create(ActorInitializer init) { return new EngineerRepair(init, this); }
+		[Desc("Sound to play when repairing is done.")]
+		public readonly string RepairSound = null;
+
+		[Desc("Cursor to show when hovering over a valid actor to repair.")]
+		public readonly string Cursor = "goldwrench";
+
+		[Desc("Cursor to show when target actor has full health so it can't be repaired.")]
+		public readonly string FullHealthCursor = "goldwrench-blocked";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: Please call this `RepairBlockedCursor` for consistency with our other trait cursor fields.

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 4]
```diff
@@ -80,46 +95,54 @@ public void ResolveOrder(Actor self, Order order)
...
+			public EngineerRepairOrderTargeter(EngineerRepairInfo info)
+				: base("EngineerRepair", 6, info.Cursor, true, true)
+			{
+				this.info = info;
+			}
 
 			public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
 			{
-				if (!target.Info.HasTraitInfo<EngineerRepairableInfo>())
+				var engiRepairable = target.Info.TraitInfoOrDefault<EngineerRepairableInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: `var engineerRepairable`

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 5]
```diff
@@ -80,46 +95,54 @@ public void ResolveOrder(Actor self, Order order)
...
+				if (engiRepairable == null)
+					return false;
+
+				if (info.Types.Any() && engiRepairable.Types.Any() && !engiRepairable.Types.Overlaps(info.Types))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: In all other cases we only check the base bitset value, and we prefer `.IsEmpty` over `.Any()` in bitsets.

So this line should be `if (!engineerRepairable.Type.IsEmpty && !engineerRepairable.Type(info.Types))`

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 6]
```diff
@@ -80,46 +95,54 @@ public void ResolveOrder(Actor self, Order order)
...
-					cursor = "goldwrench-blocked";
+					cursor = info.FullHealthCursor;
 
 				return true;
 			}
 
 			public override bool CanTargetFrozenActor(Actor self, FrozenActor target, TargetModifiers modifiers, ref string cursor)
 			{
-				if (!target.Info.HasTraitInfo<EngineerRepairableInfo>())
+				var engiRepairable = target.Info.TraitInfoOrDefault<EngineerRepairableInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: `engineerRepairable`

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 7]
```diff
@@ -80,46 +95,54 @@ public void ResolveOrder(Actor self, Order order)
...
+				if (engiRepairable == null)
 					return false;
 
-				if (self.Owner.Stances[target.Owner] != Stance.Ally)
+				if (info.Types.Any() && engiRepairable.Types.Any() && !engiRepairable.Types.Overlaps(info.Types))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: `if (!engineerRepairable.Type.IsEmpty && !engineerRepairable.Type(info.Types))`

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs (hunk 2 above)
Comment (MustaphaTR) [reply OpenRA.Mods.Common/Traits/EngineerRepair.cs]: Copy pasted from Cloneable, missed that part.

**Comment by MustaphaTR** [comment]
Updated.

**Review by pchote — DISMISSED** [review]
I assume you already know about the conflict between this and capturing, and that it doesn't matter for your usecase here.

Code changes look reasonable and this doesn't appear to regress anything in the default mods, so LGTM.

I'm assuming of course that the testcase will be removed before merging.

File: OpenRA.Mods.Common/Traits/EngineerRepairable.cs [hunk 8]
```diff
@@ -9,12 +9,19 @@
...
 
 namespace OpenRA.Mods.Common.Traits
 {
+	public class EngineerRepairType { }
+
 	[Desc("Eligible for instant repair.")]
-	class EngineerRepairableInfo : TraitInfo<EngineerRepairable> { }
+	class EngineerRepairableInfo : TraitInfo<EngineerRepairable>
+	{
+		[Desc("Actors with these Types under EngineerRepair trait which can repair me.")]
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/EngineerRepairable.cs]: Remove the 'which'.

**Comment by MustaphaTR** [comment]
Updated.

**Comment by reaperrr** [comment]
👍 after testcase is removed.

**Comment by MustaphaTR** [comment]
Removed testcase.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ef9bcde0a194471dc20ec778beff73b0a09177e4...8d18f310c1bf0419a8b15545beb75965d3de6c58

**Pull request merged**